common/beerocks/bcl/include/bcl/network/socket_refactorings.h

```diff
@@ -0,0 +1,795 @@
+    virtual int fd()          = 0;
+    virtual bool close()      = 0;
+};
+
+/**
+ * File descriptor implementation.
+ * This class is basically a wrapper around an `int fd`
+ * This class will be aggregated from all classes modeling OS resources implementing the file
+ * descriptor interface.
+ */
+class FileDescriptorImpl : public FileDescriptor {
+public:
+    explicit FileDescriptorImpl(int fd) : m_fd(fd)
+    {
+        // TODO: error if FileDescriptor::invalid_descriptor == fd;
+    }
+
+    /**
+     * Delete copy constructor to avoid having multiple copies of the file descriptor
+     */
+    FileDescriptorImpl(const FileDescriptorImpl &) = delete;
+
+    /**
+     * Delete assignment operator to avoid having multiple copies of the file descriptor
+     */
+    FileDescriptorImpl &operator=(const FileDescriptorImpl &) = delete;
+
+    /**
+     * A move constructor allows the resources owned by an rvalue object to be moved into an
+     * lvalue without creating its copy.
+     */
+    FileDescriptorImpl(FileDescriptorImpl &&obj)
+    {
+        this->m_fd = obj.m_fd;
+        obj.m_fd   = FileDescriptor::invalid_descriptor;
+    }
+
+    /**
+     * Close socket on destructor (if it is still open)
+     */
+    ~FileDescriptorImpl() override { close(); }
+
+    /**
+     * @brief Gets file descriptor.
+     *
+     * File descriptor can be used:
+     * - in select(), poll() or epoll() to wait for events on this descriptor.
+     * - in send() family of functions to send data through the socket connection.
+     * - in recv() family of functions to receive data from the socket connection.
+     *
+     * @return File descriptor value.
+     */
+    int fd() override { return m_fd; }
+
+    /**
+     * @brief Closes file descriptor.
+     *
+     * If valid, closes file descriptor and then invalidates it.
+     *
+     * @return True on success and false otherwise (for example, if it was already closed)
+     */
+    bool close() override
+    {
+        printf("close(%d)\n", m_fd);
+        if (FileDescriptor::invalid_descriptor == m_fd) {
+            return false;
+        }
+
+        int rc = ::close(m_fd);
+        m_fd   = FileDescriptor::invalid_descriptor;
+
+        return (0 == rc);
+    }
+
+private:
+    int m_fd;
+};
+
+/**
+ * Wrapper class around sockadr (structure describing a generic socket address)
+ */
+class SockAddr {
+public:
+    virtual ~SockAddr()                             = default;
+    virtual const struct sockaddr *sockaddr() const = 0;
+    virtual const socklen_t &length() const         = 0;
+
+    struct sockaddr *sockaddr()
+    {
+        return const_cast<struct sockaddr *>(const_cast<const SockAddr *>(this)->sockaddr());
+    }
+    socklen_t &length()
+    {
+        return const_cast<socklen_t &>(const_cast<const SockAddr *>(this)->length());
+    }
+};
+
+/**
+ * Wrapper class around msghdr (structure describing messages sent by `sendmsg` and received by
+ * `recvmsg`)
+ */
+class MsgHdr {

```

Comment:
I think that creating a class that its sole propose is to be an opt-in to some C structure is meaningless.
This is the structure - 
```
struct msghdr {
    void         *msg_name;       /* optional address */
    socklen_t     msg_namelen;    /* size of address */
    struct iovec *msg_iov;        /* scatter/gather array */
    size_t        msg_iovlen;     /* # elements in msg_iov */
    void         *msg_control;    /* ancillary data, see below */
    size_t        msg_controllen; /* ancillary data buffer len */
    int           msg_flags;      /* flags on received message */
};
```
When I think of wrapping it with a C++ class I imagine that I implement the C++ class with _my specific needs_ and adding a layer of simplicity and a layer of safety. Otherwise I don't think there is a need to wrap, it just adds additional  complexity.

Reply:
I agree. As @arnout has pointed out [here](https://github.com/prplfoundation/prplMesh/pull/1499/files#r449661884), I'll probably remove this class and all the calls to `recvmsg()`